### PR TITLE
US77579: Created an interface that connectors should implement to get…

### PR DESCRIPTION
**Pull request by anubhavi25**
… attributes and refactored existing code to adhere to said interface

Signed-off-by: Irina <irina.epshteyn1@ge.com>

### CODE REVIEW

**Comment by anubhavi25** [comment]
[Passing Builds: 111-113,116](https://predix1.jenkins.build.ge.com/job/Predix-Security/job/ACS/job/ACS-PullRequest/job/acs-integration-pullrequest/)

File: service/src/main/java/com/ge/predix/acs/attribute/connectors/AttributeStore.java [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package com.ge.predix.acs.attribute.connectors;
+
+import com.ge.predix.acs.model.Attribute;
+
+import java.util.Set;
+
+public interface AttributeStore {
+    Set<Attribute> getAttributes(String identifier);
```
Comment (FrankGasparovic) [inline service/src/main/java/com/ge/predix/acs/attribute/connectors/AttributeStore.java]: Extend this interface to DefaultAttributeStore with method Set<Attribute> getAttributesByCriteria(String identifier, Set<Attributes> criteria)

File: service/src/test/java/com/ge/predix/acs/attribute/connectors/DefaultSubjectAttributeStoreTest.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    @Test(expectedExceptions = ClassCastException.class)
+    public void testGetAttributesByInvalidCriteria() throws Exception {
+        BaseSubject subject = new BaseSubject("someSubject");
+        Mockito.when(this.privilegeManagementService.getBySubjectIdentifierAndScopes(Mockito.any(), Mockito.any()))
+               .thenReturn(subject);
+        Set<String> criteria = Collections.singleton("someCriteria");
+        this.defaultSubjectAttributeStore.getAttributesByCriteria(subject.getSubjectIdentifier(), criteria);
+    }
```
Comment (FrankGasparovic) [inline service/src/test/java/com/ge/predix/acs/attribute/connectors/DefaultSubjectAttributeStoreTest.java]: This expectedException would change with the new DefaultAttributeStore interface. See above comment

File: service/src/test/java/com/ge/predix/acs/service/policy/evaluation/PolicyEvaluationWithAttributeUriTemplateTest.java [hunk 3]
```diff
@@ -71,11 +73,11 @@
     @Mock
     private PolicyEvaluationCacheCircuitBreaker cache; 
 
-    @SuppressWarnings("unchecked")
     @Test(enabled = true)
```
Comment (FrankGasparovic) [inline service/src/test/java/com/ge/predix/acs/service/policy/evaluation/PolicyEvaluationWithAttributeUriTemplateTest.java]: Why explicitly enable the test?

File: service/src/main/java/com/ge/predix/acs/attribute/connectors/AttributeStore.java [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+    Set<Attribute> getAttributesByCriteria(String identifier, Set<?> criteria);
```
Comment (sanjeevchopra) [inline service/src/main/java/com/ge/predix/acs/attribute/connectors/AttributeStore.java]: * getting subject attributes by 'scope' has a specific meaning, so i suggest we keep that naming.   `getAttributesByScope(String identified, Set<Attribute> scopes)` .  
* keep the criteria to a specific type  (`Set<Attribute>`)  and name(scopes) as well. When we have another usecase, we can add a new method for it.
* also raises the question if we should have 2 separate interfaces at this level for Subject and Resource - the underlying implementation can still be shared.

File: service/src/main/java/com/ge/predix/acs/attribute/connectors/AttributeStore.java [hunk 5]
```diff
@@ -0,0 +1,10 @@
+package com.ge.predix.acs.attribute.connectors;
+
+import com.ge.predix.acs.model.Attribute;
+
+import java.util.Set;
+
+public interface AttributeStore {
```
Comment (sanjeevchopra) [inline service/src/main/java/com/ge/predix/acs/attribute/connectors/AttributeStore.java]: maybe use a name which implies that is is read-only. AttributeReader or similar ?

File: service/src/test/java/com/ge/predix/acs/service/policy/evaluation/SubjectAttributeResolverTest.java [hunk 6]
```diff
@@ -115,36 +106,11 @@ public void testGetSubjectAttributesSubjectFoundWithAttributesAndNoSupplementalA
 
     @Test
     public void testGetSubjectAttributesSupplementalAttributesOnly() {
-        BaseSubject testSubject = new BaseSubject();
-        testSubject.setSubjectIdentifier("/test/subject");
-
         Set<Attribute> supplementalSubjectAttributes = new HashSet<>();
         supplementalSubjectAttributes.add(new Attribute("https://acs.attributes.int", "site", "sanramon"));
```
Comment (FrankGasparovic) [inline service/src/test/java/com/ge/predix/acs/service/policy/evaluation/SubjectAttributeResolverTest.java:111]: Looks like this test isn't asserting anything nor expecting any exceptions?

File: service/src/main/java/com/ge/predix/acs/service/policy/evaluation/ResourceAttributeResolver.java [hunk 7]
```diff
@@ -44,9 +42,10 @@
      * @param requestResourceUri
      *            URI of the resource from the policy evaluation request
      */
-    public ResourceAttributeResolver(final PrivilegeManagementService privilegeService, final String requestResourceUri,
-            final Set<Attribute> supplementalResourceAttributes) {
-        this.privilegeService = privilegeService;
+    public ResourceAttributeResolver(final AttributeStore resourceAttributeStore,
+                                     final String requestResourceUri,
+                                     final Set<Attribute> supplementalResourceAttributes) {
```
Comment (sanjeevchopra) [inline service/src/main/java/com/ge/predix/acs/service/policy/evaluation/ResourceAttributeResolver.java]: this format is better, though is not per our current project settings. pls re-concile

File: service/src/test/java/com/ge/predix/acs/attribute/connectors/DefaultResourceAttributeStoreTest.java [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    @Autowired
+    @InjectMocks
+    private DefaultResourceAttributeStore defaultResourceAttributeStore;
+
+    @BeforeMethod
+    public void beforeMethod() {
+        MockitoAnnotations.initMocks(this);
+    }
+
+    @Test
```
Comment (sanjeevchopra) [inline service/src/test/java/com/ge/predix/acs/attribute/connectors/DefaultResourceAttributeStoreTest.java]: pls add a test for existing resource as well.  (and subject for DefaultSubject store)

File: service/src/test/java/com/ge/predix/acs/attribute/connectors/DefaultSubjectAttributeStoreTest.java [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+    @Test
+    public void testGetAttributesForNonExistentSubject() throws Exception {
+        Mockito.when(this.privilegeManagementService.getBySubjectIdentifierWithInheritedAttributes(Mockito.any()))
+               .thenReturn(null);
+        Assert.assertTrue(this.defaultSubjectAttributeStore.getAttributes("nonexistentSubject").isEmpty());
+    }
+
+    @Test
+    public void testGetAttributesByCriteriaForNonExistentSubject() throws Exception {
```
Comment (sanjeevchopra) [inline service/src/test/java/com/ge/predix/acs/attribute/connectors/DefaultSubjectAttributeStoreTest.java]: for all criteria related test, isn't it sufficient to test that DefaultSubjectAttributeStore calls the corresponding call on the PrivilegeManagementService with the appropriate criteria/scopes ?

File: service/src/test/java/com/ge/predix/acs/service/policy/evaluation/ResourceAttributeResolverTest.java [hunk 10]
```diff
@@ -70,44 +77,87 @@ public void testResolveResourceUri(final String resourceURI, final String attrib
...
+    public void testGetResourceAttributesResourceFoundWithNoAttributesAndNoSupplementalAttributes() {
+        ResourceAttributeResolver resolver = new ResourceAttributeResolver(this.defaultResourceAttributeStore,
+                this.testResource.getResourceIdentifier(), null);
+
+        Set<Attribute> combinedResourceAttributes = resolver.getResult(getPolicy(null)).getResourceAttributes();
+        Assert.assertTrue(combinedResourceAttributes.isEmpty());
+    }
+
+    @Test
+    public void testGetSubjectAttributesSubjectFoundWithAttributesAndNoSupplementalAttributes() {
```
Comment (sanjeevchopra) [inline service/src/test/java/com/ge/predix/acs/service/policy/evaluation/ResourceAttributeResolverTest.java]: testGetResourceAttributesWithAttributesAndNoSupplementalAttributes ?

**Review by sanjeevchopra — DISMISSED** [review]

File: service/src/test/java/com/ge/predix/acs/service/policy/evaluation/SubjectAttributeResolverTest.java (hunk 6 above)
Comment (irinaepshteyn) [reply service/src/test/java/com/ge/predix/acs/service/policy/evaluation/SubjectAttributeResolverTest.java:111]: @FrankGasparovic , this test is asserting that if subject doesn't have any attributes but supplemental attributes are passed in, then supplemental attributes are returned by the attribute resolver.

**Review by sanjeevchopra — APPROVED** [review]

**Pull request merged**